### ISSUE TITLE
Suggestion: Reaper Transport State as Input

### ISSUE THREAD

**Issue by AtmanActive** [helgoboss/helgobox#634]
Hi,

Congratulations for writing this beautiful software, it is amazing indeed. It takes Reaper to new heights.

Sometimes in studio work, during complex recording sessions there is a need to automate certain track/FX/send parameters depending on Reaper's transport state (stop/play/rec), and since ReaLearn is already so powerful, this would be the right addition to its arsenal.
Reaper Transport State as input, right next to MIDI, OSC and keyboard.

Thanks.

Cheers!

**Comment by AtmanActive** [comment]
If you ever consider this, I have one more suggestion: for ease of rule programming, the events should have separate play-but-not-recording and play-and-recording.

So the full list could be:

 - Reaper: Transport: Stop
 - Reaper: Transport: Play
 - Reaper: Transport: Record
 - Reaper: Transport: Play-or-Record
 - Reaper: Transport: Pause
 - Reaper: Transport: Unpause

**Comment by helgoboss** [comment]
Hi @AtmanActive. Sorry, I thought I already replied to this one but it doesn't seem so :)

I'm considering to implement this in a more generic way one day. You basically want to use one of ReaLearn's *targets* ("Project: Invoke transport action") as a *source*. Maybe I can add a source type one day that "borrows" the target of another mapping and uses it as a source. It would be similar to how target-based conditional activation is currently implemented: In the "Active" dropdown on the top of the mapping panel, you can choose "When target value met" and there pick another mapping.

**Comment by AtmanActive** [comment]
Yes, that's the idea: control targets when Reaper changes it's transport state.
Like, for example:
- bypass an FX when recording, unbypass on stop
- mute a track on playback, unmute on stop
- light up a MIDI red light on record, turn it off on stop
...

Thanks!

**Comment by helgoboss** [comment]
> * light up a MIDI red light on record, turn it off on stop

The other examples are fine. But this one can be achieved in a much better way by using the existing feedback functionality of ReaLearn.

**Comment by AtmanActive** [comment]
Interesting.

Care to elaborate?

**Comment by helgoboss** [comment]
Sure! If you have a controller with LED buttons, you should set "Feedback output" to that controller. Then (if the controller is "nice"), a mapping of such an LED button to "Project: Invoke transport action" with "Record" should just give you what you want: The LED should light up as long as the project is recording. That's what "feedback" is all about in ReaLearn.

**Comment by MonkeyBars3k** [comment]
> I'm considering to implement this in a more generic way one day. You basically want to use one of ReaLearn's _targets_ ("Project: Invoke transport action") as a _source_. Maybe I can add a source type one day that "borrows" the target of another mapping and uses it as a source. It would be similar to how target-based conditional activation is currently implemented: In the "Active" dropdown on the top of the mapping panel, you can choose "When target value met" and there pick another mapping. This _existing_ feature lets you enable/disable mappings depending on if the project is playing/recording/etc. The _new_ feature would also let you control targets whenever the transport state changes.

Yes, that would be excellent. However, if possible, it would be even better if this new feature didn't require another mapping to create (or perhaps both approaches could be added). I.e. some sort of source selection that grabs the data internally from one of the target options...

**Comment by MonkeyBars3k** [comment]
Hi there! Any thoughts on this FR? You mentioned in the forum that it wouldn't be difficult to implement. The uses are endless – it would make ReaLearn SO much more powerful!

**Comment by helgoboss** [comment]
Not very difficult, but can't do it at the moment due to a lack of time (full time job + Playtime). It will come eventually...

**Comment by mccrabney** [comment]
`Maybe I can add a source type one day that "borrows" the target of another mapping and uses it as a source.`

+1, this one would unlock an impressive realm of possibilities!

**Comment by MonkeyBars3k** [comment]
As an extra bit of motivation, I will reveal that I'm working on a script that will enable a long-desired routing capability in Reaper that depends on this enhancement. So, by doing it your work will be magnified by my project :)

**Comment by MonkeyBars3k** [comment]
Does the recent huge revamp and upgrade (congratulations!) make this FR any easier?

**Comment by helgoboss** [comment]
Thanks. No, nothing happened in that direction. I guess some funding would make this FR easier ;)

**Comment by AtmanActive** [comment]
> I guess some funding would make this FR easier ;)

How can we donate so that the donation goes toward this development please?

**Comment by helgoboss** [comment]
Good question. I'm thinking about setting up a possibility to donate towards particular features, but there's nothing there yet.   Probably using a popular platform such as Patreon (or something similar, haven't looked yet into the differences). Stay tuned.

**Comment by AtmanActive** [comment]
https://oss.issuehunt.io/

**Comment by helgoboss** [comment]
@AtmanActive Nice, thanks for the suggestion. Will look into it.
